canopen_402/src/motor.cpp

```diff
@@ -352,22 +325,46 @@ bool Motor402::switchMode(LayerStatus &status, uint16_t mode) {
     if(!switchState(status, State402::Operation_Enable)) return false;
 
     return okay;
-
 }
+State402::InternalState Motor402::nextStateForEnabling(State402::InternalState state){
+    switch(state){
+    case State402::Start:
+        return State402::Not_Ready_To_Switch_On;
+
+    case State402::Fault:
+    case State402::Not_Ready_To_Switch_On:
+        return State402::Switch_On_Disabled;
+
+    case State402::Switch_On_Disabled:
+        return State402::Ready_To_Switch_On;
+
+    case State402::Ready_To_Switch_On:
+        return State402::Switched_On;
+
+    case State402::Switched_On:
+    case State402::Quick_Stop_Active:
+    case State402::Operation_Enable:
+        return State402::Operation_Enable;
 
+    case State402::Fault_Reaction_Active:
+        return State402::Fault;
+    }
+}
 bool Motor402::switchState(LayerStatus &status, const State402::InternalState &target){
     time_point abstime = get_abs_time(state_switch_timeout_);
     State402::InternalState state = state_handler_.getState();
     target_state_ = target;
     while(state != target_state_){
         boost::mutex::scoped_lock lock(cw_mutex_);
-        State402::InternalState next = State402::Unknown;
-        if(!Command402::setTransition(control_word_ ,state, target_state_ , &next)){
+        State402::InternalState hop = target_state_;
+        if(target_state_ == State402::Operation_Enable || target_state_ == no_mode_state_)
+            hop = nextStateForEnabling(state);
```

Comment:
This might not be possible for all cases..

Reply:
Let's have a look at the power state machine.

IMO the following states are possible "no mode states":

1. Switch on disabled
2. Ready to switch on
3. Switched on
4. Operated enabled

Should we check that and add an exception?

If switching upwards (towards Operation enabled) then "nextStateForEnabling" should return the correct next state. It's just like switching to Operation enabled but will succeed earlier.

However when switching downwards then "nextStateForEnabling" is not correct. If not in fault or quick stop we can switch to the downward state directly. If in fault or quick stop "nextStateForEnabling" should work again.

The difficulty here is to determine when to use nextStateForEnabling. IMO this information is present in the transistion table. So I'd use nextStateForEnabling whenever setTransistion returns false. Let me add that ...